sycl/source/detail/async_alloc.cpp

```diff
@@ -0,0 +1,347 @@
+  return impl->getPropList();
+}
+
+__SYCL_EXPORT memory_pool::memory_pool(const sycl::context &ctx,
+                                       const sycl::device &dev,
+                                       const sycl::usm::alloc kind,
+                                       const property_list &props) {
+
+  if (kind == sycl::usm::alloc::host) {
+    throw sycl::exception(
+        sycl::make_error_code(sycl::errc::invalid),
+        "Host allocated memory pools selected but device supplied!");
+  }
+
+  if (kind != sycl::usm::alloc::device) {
+    throw sycl::exception(
+        sycl::make_error_code(sycl::errc::feature_not_supported),
+        "Only device allocated memory pools are supported!");
+  }
+
+  impl = std::make_shared<detail::memory_pool_impl>(ctx, dev, kind, props);
+}
+
+// NOT SUPPORTED: Host side pools unsupported.
+__SYCL_EXPORT memory_pool::memory_pool(const sycl::context &,
+                                       const property_list &) {
+  throw sycl::exception(
+      sycl::make_error_code(sycl::errc::feature_not_supported),
+      "Host allocated pools are unsupported!");
+}
+
+__SYCL_EXPORT
+memory_pool::memory_pool(const sycl::queue &q, const sycl::usm::alloc kind,
+                         const property_list &props)
+    : memory_pool(q.get_context(), q.get_device(), kind, props) {}
+
+// NOT SUPPORTED: Creating a pool from an existing allocation is unsupported.
+__SYCL_EXPORT memory_pool::memory_pool(const sycl::context &, const void *,
+                                       size_t, const property_list &) {
+  throw sycl::exception(
+      sycl::make_error_code(sycl::errc::feature_not_supported),
+      "Creating a pool from an existing allocation is unsupported!");
+}
+
+// <--- Async allocs/frees --->
+__SYCL_EXPORT void *async_malloc(sycl::handler &h, sycl::usm::alloc kind,
+                                 size_t size) {
+
+  // Non-device allocations are unsupported.
+  if (kind != sycl::usm::alloc::device) {
+    throw sycl::exception(
+        sycl::make_error_code(sycl::errc::feature_not_supported),
+        "Host allocated pools are unsupported!");
+  }
+
+  auto &Adapter = h.getContextImplPtr()->getAdapter();
+  auto &Q = h.MQueue->getHandleRef();
+
+  // Get events to wait on.
+  auto depEvents = detail::getUrEvents(h.impl->CGData.MEvents);
+  uint32_t numEvents = h.impl->CGData.MEvents.size();
+
+  void *alloc = nullptr;
+  ur_event_handle_t Event;
+  Adapter->call<sycl::errc::runtime,
+                sycl::detail::UrApiKind::urEnqueueUSMDeviceAllocExp>(
+      Q, (ur_usm_pool_handle_t)0, size, nullptr, numEvents, depEvents.data(),
+      &alloc, &Event);
+
+  // Async malloc must return a void* immediately.
+  // Set up CommandGroup which is a no-op and pass the event from the alloc.
+  h.impl->MAllocSize = size;
+  h.impl->MAsyncAllocEvent = Event;
+  h.setType(detail::CGType::AsyncAlloc);
+
+  return alloc;
+}
+
+__SYCL_EXPORT void *async_malloc(const sycl::queue &q, sycl::usm::alloc kind,
```

Comment:
This will (probably) merge before we add support for graphs, so we should throw an error if a user tries that. If you look at `sycl/source/handler.cpp`  you will see calls to `throwIfGraphAssociated<>`. That's now we handle using unsupported handler methods in a graph at the moment.

This should be done for all the new entry-points, with unittests in `sycl/unittests/Extensions/CommandGraph/Exceptions.cpp` to verify that's working. Although you might need to uncomment https://github.com/intel/llvm/blob/sycl/sycl/unittests/Extensions/CommandGraph/CMakeLists.txt#L3C2-L3C45 during development of those tests.

Reply:
nice one, thanks